packages/core/plugins/core-plugins-contracts-server/src/contracts.ts

```diff
@@ -0,0 +1,184 @@
+ * Side Public License, v 1.
+ */
+
+import type { PluginName } from '@kbn/core-base-common';
+
+/**
+ * Setup contract of Core's `plugins` service.
+ *
+ * @public
+ */
+export interface PluginsServiceSetup {
+  /**
+   * Returns a promise that will resolve with the requested plugin setup contracts once all plugins have been set up.
+   *
+   * If called when plugins are already setup, the returned promise will resolve instantly.
+   *
+   * The API can only be used to resolve required dependencies, optional dependencies, or dependencies explicitly
+   * defined as `runtimePluginDependencies` in the calling plugin's manifest, otherwise the API will throw at call time.
+   *
+   * **Important:** This API should only be used when trying to address cyclic dependency issues that can't easily
+   * be solved otherwise. This is meant to be a temporary workaround only supposed to be used until a better solution
+   * is made available.
+   * Therefore, by using this API, you implicitly agree to:
+   * - consider it as technical debt and open an issue to track the tech debt resolution
+   * - accept that this is only a temporary solution, and will comply to switching to the long term solution when asked by the Core team
+   *
+   * @remark The execution order is not guaranteed to be consistent. Only guarantee is that the returned promise will be
+   *         resolved once all plugins are started, and before Core's `start` lifecycle is resumed.
+   *
+   * @example
+   * ```ts
+   * setup(core) {
+   *   core.plugins.onSetup<{pluginA: SetupContractA, pluginB: SetupContractA}>('pluginA', 'pluginB')
+   *       .then(({ pluginA, pluginB }) => {
+   *         if(pluginA.found && pluginB.found) {
+   *           // do something with pluginA.contract and pluginB.contract
+   *         }
+   *       });
+   * }
+   *
+   * @experimental
+   * ```
+   */
+  onSetup: PluginContractResolver;
+  /**
+   * Returns a promise that will resolve with the requested plugin start contracts once all plugins have been started.
+   *
+   * If called when plugins are already started, the returned promise will resolve instantly.
+   *
+   * The API can only be used to resolve required dependencies, optional dependencies, or dependencies explicitly
+   * defined as `runtimePluginDependencies` in the calling plugin's manifest, otherwise the API will throw at call time.
+   *
+   * **Important:** This API should only be used when trying to address cyclic dependency issues that can't easily
+   * be solved otherwise. This is meant to be a temporary workaround only supposed to be used until a better solution
+   * is made available.
+   * Therefore, by using this API, you implicitly agree to:
+   * - consider it as technical debt and open an issue to track the tech debt resolution
+   * - accept that this is only a temporary solution, and will comply to switching to the long term solution when asked by the Core team
```

Comment:
The actual disclaimer, present on all the APIs, both server and client side.

Reply:
I think we should add a note that runtime depedencies will not always resolve (plugins could be disabled) and code should ensure there's a smooth user experience for both found/!found.

To avoid copy-pasta it might be good to also update the example code with an explicit `else`